File: techila/examples/CSharp/Tutorial/3_datafiles/RunDatafiles.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Techila.Management;
using System.Threading;

namespace RunDatafiles
{
    class RunDatafiles
    {
        static void Main()
        {
            // Will be used to define the number of Jobs
            int jobs = 4;

            // Will be used to define the size of the random number array
            int arraySize = 5;

            // Create text files containing sample values
            for (int i = 0; i < jobs; i++) // For each file
            {
                // Create reproducible randoms
                Random rand = new Random(i);

                // Array for storing random numbers
                int[] rngArray = new int[arraySize];

                // Define name of file that will created
                string fname = "file" + i + ".txt";

                // Create object for writing to file
                System.IO.StreamWriter file = new System.IO.StreamWriter(fname);

                // Generate 5 random values and store in file
                for (int j = 0; j < arraySize; j++)
                {
                    int rngValue = rand.Next(0, 101);
                    string rngValueStr = rngValue.ToString();
                    file.WriteLine(rngValueStr);
                }

                // Close the file
                file.Close();
            }

            // Create TechilaManager instance
            TechilaManager tm = TechilaManagerFactory.TechilaManager;

            // Test the connection to the Techila Server by creating a session.
            // Returns value 0 if session was created successfully.
            int status = tm.InitFile();

            // Create Support instance
            Support sp = tm.Support();

            // Get the status code description.
            string codedesc = sp.DescribeStatusCode(status);

            // Print the status code description
            Console.WriteLine("Status: " + codedesc);

            // Create a TechilaProject instance and link it with the 'tm' object created earlier
            TechilaProject tp = new TechilaProject(tm, "Datafiles Tutorial");

            // Enable messages
            tp.GetPeach().Messages = true;

            // Create Bundle for transferring files to Workers
            tp.GetPeach().NewDataBundle();

            // Add following 4 files to the Bundle
            tp.GetPeach().AddDataFile("file0.txt");
            tp.GetPeach().AddDataFile("file1.txt");
            tp.GetPeach().AddDataFile("file2.txt");
            tp.GetPeach().AddDataFile("file3.txt");

            // Create 5 jobs, each job will execute the class with a different input argument
            for (int i = 0; i < jobs; i++)
            {
                tp.Add(new DatafilesDist(i));
            }

            // Create and start the Project. Execution will continue after all Jobs have been completed.
            // Job results will be stored in the 'tp' object.
            tp.Execute();

            for (int i = 0; i < jobs; i++)
            {
                // Retrieve one Job result from 'tp'
                int jobresult = ((DatafilesDist)tp.Get(i)).Result;

                // Display the value generated in the Job
                Console.WriteLine("Result of Job {0}: {1}", i + 1, jobresult);
            }

            // Uninitialize and remove the session
            tm.Unload(true);
        }

        // Mark the class as serializable
        [Serializable]
        class DatafilesDist : TechilaThread // Extend the TechilaThread class
        {
            // Define type of input argument
            int Jobidx;

            // Define 'Result' as public so it can be accessed when Job results are retrieved from the 'tp' list.
            public int Result;

            // Define constructor that takes one input argument
            public DatafilesDist(int jobidx)
            {
                this.Jobidx = jobidx;
            }

            // Override the 'Execute' method in the 'TechilaThread' class with code that we want to execute
            public override void Execute()
            {
                // Choose file that will be read during one Job
                string fname = "file" + Jobidx + ".txt";
                string line;

                // Initialize the result counter to 0.
                Result = 0;

                // Create object for reading from file
                System.IO.StreamReader file = new System.IO.StreamReader(fname);

                while ((line = file.ReadLine()) != null) // Sum values on lines
                {
                    // Increase summation result by the value of one random value read from file
                    Result = Result + Convert.ToInt32(line);
                }

                // Close the file
                file.Close();

            }

        }

    }
}

File: techila/examples/java/tes/Tutorial/3_datafiles/RunDatafiles.java
// Copyright 2017 Techila Technologies Ltd.

import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.io.PrintWriter;
import java.io.IOException;
import java.util.Vector;

import java.util.Random;

import fi.techila.user.Support;
import fi.techila.user.TechilaExecutorCompletionService;
import fi.techila.user.TechilaExecutorService;
import fi.techila.user.TechilaFutureTask;
import fi.techila.user.TechilaManagerFactory;
import fi.techila.user.TechilaManager;

public class RunDatafiles {

    public void runDatafiles(int jobs) {
        // Method for creating the computational Project. In this
        // example, a set of input files will be transferred from your
        // computer to the Techila Workers.

        // Create a TechilaManager object, which will be used for
        // initializing the network connection / authentication
        TechilaManager tm = TechilaManagerFactory.getTechilaManager();

        try {
            // Generate some input files
            Vector<String> fileList = new Vector<String>();
            Random rand = new Random();
            for (int i = 0;i < jobs;i++) {
                try{
                    // Each file will be given a unique name.
                    String filename = "TechilaExampleFile"
                        + Integer.toString(i) + ".txt";
                    PrintWriter filewriter = new PrintWriter(filename, "UTF-8");
                    int randomNum;
                    // Each file will contain five rows, each row will
                    // contain a random number.
                    for (int j=0; j < 6; j++) {
                        randomNum = rand.nextInt(101);
                        filewriter.println(Integer.toString(randomNum));
                    }
                    filewriter.close();

                    // Store file name to 'fileList'
                    fileList.add(filename);
                } catch (IOException ee) {
                    System.err.println("Exception: " + ee);
                    ee.printStackTrace();
                }
            }

            // Initialize a connection to the Techila Server
            int status = tm.initFile();

            // Print the status of initialization attempt
            Support sp = tm.support();
            String codedesc = sp.describeStatusCode(status);
            System.out.println("Status: " + codedesc);

            // Create required TES objects.
            TechilaExecutorService<DatafilesDist> tes =
                new TechilaExecutorService<DatafilesDist>(tm,
                                                          "DatafilesTutorial");
            TechilaExecutorCompletionService<DatafilesDist, Integer> cs =
                new TechilaExecutorCompletionService<DatafilesDist, Integer>(tes);

            // Enable messages.
            tes.getPeach().setMessages(true);

            // Create Bundle for transferring files to Techila Workers
            tes.getPeach().newDataBundle();

            // Add files to the Bundle
            for (String fname : fileList) {
                tes.getPeach().addDataFile(fname);
            }

            // Create Jobs. Each DatafilesDist object will be
            // contructed by a different input argument. This input
            // argument will be used to determine which input file
            // should be processed.
            for (int i = 0; i < jobs; i++) {
                cs.submit(new DatafilesDist(i));
            }

            // Start the computational Project. Also starts a local
            // thread that will automatically download the results as
            // soon as they are available.
            tes.start();

            // Access and print the results.
            for (int i = 0; i < jobs; i++) {
                try {
                    TechilaFutureTask<Integer> f = cs.take();

                    int index = f.getIndex();
                    int jobresult = f.get();

                    System.out.println("Result of summation at index "
                                       + index + ": " + jobresult);
                } catch (ExecutionException ee) {
                    System.err.println("Task failed: " + ee.getMessage());
                }
            }

            // Allow 10 seconds for the local thread to terminate.
            tes.awaitTermination(10, TimeUnit.SECONDS);

        } catch (Exception e) {
            System.err.println("Exception: " + e);
            e.printStackTrace();
        } finally {
            tm.unload(true, true);
        }
    }

    public void run() {
        runDatafiles(4);
    }

    public static void main(String[] args) {
        RunDatafiles rd = new RunDatafiles();
        rd.run();
    }
}
